host/class/uac/usb_host_uac/uac_host.c

```diff
@@ -0,0 +1,2114 @@
+        UAC_GOTO_ON_ERROR(uac_host_interface_release_and_free_transfer(iface), "Unable to release UAC Interface");
+    }
+
+    uac_host_interface_unlock(iface);
+    return ESP_OK;
+
+fail:
+    uac_host_interface_unlock(iface);
+    return ret;
+}
+
+esp_err_t uac_host_device_read(uac_host_device_handle_t uac_dev_handle, uint8_t *data, uint32_t size, uint32_t *bytes_read, uint32_t timeout)
+{
+    uac_iface_t *iface = get_iface_by_handle(uac_dev_handle);
+    UAC_RETURN_ON_INVALID_ARG(iface);
+    UAC_RETURN_ON_INVALID_ARG(data);
+    UAC_RETURN_ON_INVALID_ARG(bytes_read);
+
+    UAC_RETURN_ON_ERROR(uac_host_interface_try_lock(iface, DEFAULT_CTRL_XFER_TIMEOUT_MS), "Unable to lock UAC Interface");
+    if (UAC_INTERFACE_STATE_ACTIVE != iface->state) {
+        uac_host_interface_unlock(iface);
+        return ESP_ERR_INVALID_STATE;
+    }
+    uac_host_interface_unlock(iface);
+
+    size_t data_len = _ring_buffer_get_len(iface->ringbuf);
+    if (data_len > size) {
+        data_len = size;
+    }
+    esp_err_t ret = _ring_buffer_pop(iface->ringbuf, data, data_len, (size_t *)bytes_read, timeout);
+
+    if (ESP_OK != ret) {
+        ESP_LOGD(TAG, "RX Ringbuffer read failed");
+        return ret;
+    }
+
+    return ESP_OK;
+}
+
+esp_err_t uac_host_device_write(uac_host_device_handle_t uac_dev_handle, uint8_t *data, uint32_t size, uint32_t timeout)
+{
+    uac_iface_t *iface = get_iface_by_handle(uac_dev_handle);
+    UAC_RETURN_ON_INVALID_ARG(iface);
+    UAC_RETURN_ON_INVALID_ARG(data);
+
+    UAC_RETURN_ON_ERROR(uac_host_interface_try_lock(iface, DEFAULT_CTRL_XFER_TIMEOUT_MS), "Unable to lock UAC Interface");
+    if (UAC_INTERFACE_STATE_ACTIVE != iface->state) {
+        uac_host_interface_unlock(iface);
+        return ESP_ERR_INVALID_STATE;
+    }
+    uac_host_interface_unlock(iface);
+
+    esp_err_t ret = _ring_buffer_push(iface->ringbuf, data, size, timeout);
+
+    if (ESP_OK != ret) {
+        ESP_LOGD(TAG, "TX Ringbuffer write failed");
+        return ret;
+    }
+
+    // We need to submit the transfer if there is free transfer in the list
+    for (int i = 0; i < iface->xfer_num; i++) {
+        UAC_ENTER_CRITICAL();
+        if (iface->free_xfer_list[i]) {
+            size_t data_len = _ring_buffer_get_len(iface->ringbuf);
+            if (data_len == 0) {
+                goto exit_critical;
+            }
+            // if interface state changed to inactive during blocking write
+            // we need to return invalid state to safely exit the write function
+            if (UAC_INTERFACE_STATE_ACTIVE != iface->state) {
+                ret = ESP_ERR_INVALID_STATE;
+                goto exit_critical;
+            }
+            iface->xfer_list[i] = iface->free_xfer_list[i];
+            iface->free_xfer_list[i] = NULL;
+            iface->xfer_list[i]->status = USB_TRANSFER_STATUS_COMPLETED;
+            UAC_EXIT_CRITICAL();
+            stream_tx_xfer_done(iface->xfer_list[i]);
```

Comment:
Do I understand it right that here we start the transfer by calling transfer done callback?
If yes, I suggest modifying to logic as this seems to be counter intuitive. Or at least add a comment about what and why are we doing here

Reply:
yes, call the function is to simply keep all tranfer operate in one function.

let's split it into two functions to make it clear